File: app/src/main/java/com/inipage/homelylauncher/pager/PagerIndicatorView.java
package com.inipage.homelylauncher.pager;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.Interpolator;

import androidx.annotation.Nullable;

import com.inipage.homelylauncher.R;
import com.inipage.homelylauncher.utils.AttributeApplier;
import com.inipage.homelylauncher.utils.SizeDimenAttribute;
import com.inipage.homelylauncher.utils.SizeValAttribute;

import static com.inipage.homelylauncher.utils.AttributeApplier.intValue;

public class PagerIndicatorView extends View {

    private final int INACTIVE_ALPHA = 80;
    private final int ACTIVE_ALPHA = 220;
    private final int INDICATOR_WIDTH_FACTOR = 8;
    private final Paint mPaint;
    @SizeDimenAttribute(R.dimen.indicator_height)
    int DESIRED_HEIGHT = intValue();
    @SizeValAttribute(2)
    int STROKE_WIDTH = intValue();
    @SizeValAttribute(4)
    int INDICATOR_SIZE = intValue();
    private int mGridPageCount;
    private int mActiveItem;
    private boolean mIsSetup = false;

    public PagerIndicatorView(Context context) {
        this(context, null);
    }

    public PagerIndicatorView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public PagerIndicatorView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        AttributeApplier.applyDensity(this, getContext());
        // setBackground(context.getDrawable(R.drawable.indicator_view_background));
        mPaint = new Paint();
        mPaint.setStrokeWidth(STROKE_WIDTH);
    }

    public void setup(int gridPageCount) {
        mGridPageCount = gridPageCount;
        mActiveItem = 1;
        mIsSetup = true;
        requestLayout();
        invalidate();
    }

    public void updateActiveItem(int activeItem) {
        mActiveItem = activeItem;
        invalidate();
    }

    // Background is rendered as a View background;

    @Override
    protected void onDraw(Canvas canvas) {
        if (!mIsSetup) {
            return;
        }

        // App drawer + other pages
        final int itemCount = mGridPageCount + 1;
        final int itemWidth = getWidth() / itemCount;
        final int xOffsetInRange = (itemWidth / 2) - (INDICATOR_SIZE / 2);
        final int yOffset = (getHeight() / 2) - (INDICATOR_SIZE / 2);

        mPaint.setColor(Color.WHITE);
        for (int i = 0; i < itemCount; i++) {
            final boolean active = i == mActiveItem;
            mPaint.setAlpha(active ? ACTIVE_ALPHA : INACTIVE_ALPHA);
            mPaint.setStyle(active ? Paint.Style.FILL_AND_STROKE : Paint.Style.STROKE);
            final int xOffset = (i * itemWidth) + xOffsetInRange;
            switch (i) {
                case 0: // App drawer bubbles
                    float widthOverFour = INDICATOR_SIZE / 4F;
                    // Top left
                    canvas.drawCircle(
                        xOffset + widthOverFour,
                        yOffset + widthOverFour,
                        widthOverFour,
                        mPaint);
                    // Bottom right
                    canvas.drawCircle(
                        xOffset + (widthOverFour * 3),
                        yOffset + (widthOverFour * 3),
                        widthOverFour,
                        mPaint);
                    break;
                case 1: // Home square
                    canvas.drawRect(
                        xOffset,
                        yOffset,
                        xOffset + INDICATOR_SIZE,
                        yOffset + INDICATOR_SIZE,
                        mPaint);
                    break;
                default: // Circles
                    canvas.drawOval(
                        xOffset,
                        yOffset,
                        xOffset + INDICATOR_SIZE,
                        yOffset + INDICATOR_SIZE,
                        mPaint);
                    break;
            }
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        final int verticalSpace = MeasureSpec.getSize(heightMeasureSpec);
        setMeasuredDimension(
            mGridPageCount * INDICATOR_SIZE * INDICATOR_WIDTH_FACTOR,
            verticalSpace < DESIRED_HEIGHT ? verticalSpace : DESIRED_HEIGHT);
    }
}

File: app/src/main/java/com/inipage/homelylauncher/pager/PagerIndicatorView.kt
package com.inipage.homelylauncher.pager

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import kotlin.jvm.JvmOverloads
import com.inipage.homelylauncher.utils.SizeDimenAttribute
import com.inipage.homelylauncher.R
import com.inipage.homelylauncher.utils.AttributeApplier
import com.inipage.homelylauncher.utils.SizeValAttribute

class PagerIndicatorView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    private val paint: Paint

    @SizeValAttribute(2F)
    private var strokeWidth = AttributeApplier.floatValue()
    @SizeValAttribute(4F)
    private var rectRounding = AttributeApplier.floatValue()
    @SizeValAttribute(8F)
    private var indicatorHeight = AttributeApplier.floatValue()

    private var paddingBetweenIndicators = 0F
    private var gridPageIndicatorWidth = 0F
    private var appIndicatorRadius = 0F
    private var desiredHeight = 0F
    private var desiredWidth = 0F

    private var gridPageCount: Int = 0
    private var activePageIdx: Int = 0
    private var isSetup: Boolean = false

    fun setup(pageCount: Int) {
        gridPageCount = pageCount
        activePageIdx = 1
        desiredWidth =
            indicatorHeight +
                ((gridPageCount + 1) * paddingBetweenIndicators) +
                (gridPageCount * gridPageIndicatorWidth)
        isSetup = true
        requestLayout()
        invalidate()
    }

    fun updateActiveItem(activeItem: Int) {
        activePageIdx = activeItem
        invalidate()
    }

    // Renders as () (---) (   ) (   )
    override fun onDraw(canvas: Canvas) {
        if (!isSetup) {
            return
        }

        // We assume we have the height and width we need to layout, but we self-center inside
        // to avoid clipping the edge
        // Start at 0 and walk until the end of the view
        var xOffset = (width / 2.0F) - (desiredWidth / 2.0F)
        val yOffset = (height / 2.0F) - (desiredHeight / 2.0F)

        // App indicator
        val appIndicatorActive = 0 == activePageIdx
        paint.alpha = if (appIndicatorActive) activeAlpha else inactiveAlpha
        paint.style =
            if (appIndicatorActive) Paint.Style.FILL_AND_STROKE else Paint.Style.STROKE
        canvas.drawCircle(
            xOffset + appIndicatorRadius,
            yOffset + appIndicatorRadius,
            appIndicatorRadius,
            paint
        )
        xOffset += (appIndicatorRadius * 2.0F) + paddingBetweenIndicators

        // Page indicator(s)
        for (idx in 0 until gridPageCount) {
            val active = (idx + 1) == activePageIdx
            paint.alpha = if (active) activeAlpha else inactiveAlpha
            paint.style =
                if (active) Paint.Style.FILL_AND_STROKE else Paint.Style.STROKE
            canvas.drawRoundRect(
                xOffset,
                yOffset,
                xOffset + gridPageIndicatorWidth,
                yOffset + indicatorHeight,
                rectRounding,
                rectRounding,
                paint)
            xOffset += paddingBetweenIndicators + gridPageIndicatorWidth
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val horizontalSpace = MeasureSpec.getSize(widthMeasureSpec)
        val verticalSpace = MeasureSpec.getSize(heightMeasureSpec)
        // We self-center in the provided area
        setMeasuredDimension(horizontalSpace, verticalSpace)
    }

    init {
        AttributeApplier.applyDensity(this, getContext())
        paddingBetweenIndicators = indicatorHeight
        appIndicatorRadius = indicatorHeight / 2.0F
        gridPageIndicatorWidth = indicatorHeight * 1.5F
        desiredHeight = indicatorHeight
        paint = Paint().also {
            it.strokeWidth = strokeWidth
            it.color = Color.WHITE
        }
    }

    companion object {
        private const val inactiveAlpha = 80
        private const val activeAlpha = 255
    }
}